mrmustard/training/optimizer.py

```diff
@@ -212,13 +215,14 @@ def compute_loss_and_gradients(cost_fn: Callable, parameters: List[Parameter]):
     def should_stop(self, max_steps: int) -> bool:
         r"""Returns ``True`` if the optimization should stop (either because
         the loss is stable or because the maximum number of steps is reached)."""
+        if len(self.opt_history) < 2:
+            return False
         if max_steps != 0 and len(self.opt_history) > max_steps:
             return True
-        if len(self.opt_history) > 20:  # if cost varies less than 10e-6 over 20 steps
-            if (
-                sum(abs(self.opt_history[-i - 1] - self.opt_history[-i]) for i in range(1, 20))
-                < 1e-6
-            ):
+        self.__history_sum += abs(self.opt_history[-2] - self.opt_history[-1])
+        if len(self.opt_history) > 20:  # stop if cost varies less than 10e-6 over 20 steps
+            self.__history_sum -= abs(self.opt_history[-21] - self.opt_history[-20])
```

Comment:
I can totally be mistaken, but it seems to me here that you're only looking at... 1 element? this:
`self.opt_history[-21] - self.opt_history[-20]`
is just 1 element afaik
is that what we want to be doing here?
I'm confused when seeing this line and this comment which mentions 20 steps
`# stop if cost varies less than 10e-6 over 20 steps`
Maybe I'm just misreading

Reply:
You are misreading. We maintain the difference of the last 20 steps: in line 222 we add the last difference and if opt_history is > 20 in line 224 we remove the difference beyond 20